### ISSUE TITLE
llvm.smul.with.overflow.i64 lowered to non-existent __multi3 on PowerPC 32-bit

### ISSUE THREAD

**Issue by aaronpuchert** [llvm/llvm-project#54460]
LLVM 14 built with Clang 14 doesn't link on PowerPC (32-bit) because of
```
/usr/bin/ld: LoopStrengthReduce.cpp:(.text._ZN12_GLOBAL__N_111LSRInstance24GenerateAllReuseFormulaeEv+0xf1c): undefined reference to `__multi3'
```
The relevant lines are
```c++
    int64_t NewBaseOffset = (uint64_t)Base.BaseOffset * Factor;
    assert(Factor != 0 && "Zero factor not expected!");
    if (NewBaseOffset / Factor != Base.BaseOffset)
      continue;
```

We can build a small reproducer for that:
```c++
bool mul_overflow(unsigned long long a, long long b) {
    long long prod = a * b;
    return (prod / b != a);
}
```

Clang 13 compiles this to (with `-target powerpc-unknown-linux-gnu -O2`):
```llvm
define dso_local zeroext i1 @_Z12mul_overflowyx(i64 %0, i64 %1) local_unnamed_addr #0 {
  %3 = mul i64 %1, %0
  %4 = sdiv i64 %3, %1
  %5 = icmp ne i64 %4, %0
  ret i1 %5
}
```

So basically what it says. But Clang 14 recognizes that this is basically an overflow check and that we can get rid of the superfluous division:
```llvm
define dso_local noundef zeroext i1 @_Z12mul_overflowyx(i64 noundef %0, i64 noundef %1) local_unnamed_addr #1 {
  %3 = tail call { i64, i1 } @llvm.smul.with.overflow.i64(i64 %1, i64 %0)
  %4 = extractvalue { i64, i1 } %3, 1
  ret i1 %4
}
```
Then CodeGen produces this:
```assembly
_Z12mul_overflowyx:                     # @_Z12mul_overflowyx
.Lfunc_begin0:
        .cfi_startproc
# %bb.0:
        mflr 0
        stw 0, 4(1)
        stwu 1, -16(1)
        .cfi_def_cfa_offset 16
        .cfi_offset lr, 4
        mr      9, 3
        srawi 3, 5, 31
        srawi 7, 9, 31
        mr      10, 4
        mr      4, 3
        mr      8, 7
        bl __multi3
        srawi 5, 5, 31
        xor 3, 3, 5
        xor 4, 4, 5
        or 3, 4, 3
        cntlzw  3, 3
        not     3, 3
        rlwinm 3, 3, 27, 31, 31
        lwz 0, 20(1)
        addi 1, 1, 16
        mtlr 0
        blr
.Lfunc_end0:
        .size   _Z12mul_overflowyx, .Lfunc_end0-.Lfunc_begin0
        .cfi_endproc
```

But there is no `__multi3` on 32-bit PowerPC, because it doesn't have a 128-bit integer type. It's not in `libgcc_s.so` and compiler-rt also shouldn't have it: `compiler-rt/lib/builtins/int_types.h` has
```c
#if defined(__LP64__) || defined(__wasm__) || defined(__mips64) ||             \
    defined(__riscv) || defined(_WIN64)
#define CRT_HAS_128BIT
#endif
```
and then `compiler-rt/lib/builtins/multi3.c` is guarded with that.

**Comment by topperc** [comment]
@nickdesaulniers @LebedevRI another broken target.

**Comment by aaronpuchert** [comment]
My guess is that because `ISD::NodeType::SMULO` is not handled in `PPCDAGToDAGISel::Select`, it falls back to some generic code that expands the result to 128 bits where we can easily see whether there was an overflow. But I'm not really familiar with the backends.

If I understand this correctly, PowerPC has [overflow flags](https://devblogs.microsoft.com/oldnewthing/20180807-00/?p=99435) that should allow directly lowering the builtin. As an alternative, we might teach the middle end to stop being so clever, although it would be a (minor) loss.

**Comment by aaronpuchert** [comment]
@tstellar, feel free to move this to 14.0.1 if the release is otherwise fine, I can probably patch around this for now.

**Comment by topperc** [comment]
This has been recurring issue. Here are other patches for other targets

https://reviews.llvm.org/D112750
https://reviews.llvm.org/D108928
https://reviews.llvm.org/D108939
ttps://reviews.llvm.org/D108936
https://reviews.llvm.org/D108844
https://reviews.llvm.org/D108842

**Comment by topperc** [comment]
It used to be that llvm.smul.with.overflow would only be generated for __builtin_smul_overflow. For types that aren't legal it would use a libcall. That libcall only exists in compiler-rt and not libgcc, and it doesn't exist for all types.

In https://reviews.llvm.org/D107581 I taught the backend how to expand illegal types when the libcall was marked unavailable by the target. After that, InstCombine was changed to start using the intrinsic, but no thorough audit of the backends to disable the libcall was done. This has resulted in one target at a time being patched to fix this ever since.

**Comment by aaronpuchert** [comment]
Thanks for the insight! It seems that https://reviews.llvm.org/D108936 already added
```c++
    setLibcallName(RTLIB::MULO_I64, nullptr);
```
for 32-bit PPC. So that's good. Perhaps it should also block `RTLIB::MUL_I128`?

**Comment by aaronpuchert** [comment]
That seems to have worked out, I posted https://reviews.llvm.org/D122090. After that we get
```assembly
no__mulodi4:                            # @no__mulodi4
.Lfunc_begin0:
        .cfi_startproc
# %bb.0:                                # %entry
        stwu 1, -32(1)
        .cfi_def_cfa_offset 32
        .cfi_offset r29, -12
        .cfi_offset r30, -8
        srawi 10, 3, 31
        mulhwu 8, 3, 6
        stw 29, 20(1)                           # 4-byte Folded Spill
        li 4, 0
        stw 30, 24(1)                           # 4-byte Folded Spill
        mulhwu 11, 10, 6
        mullw 0, 3, 6
        stw 0, 0(7)
        mullw 6, 10, 6
        addc 8, 6, 8
        addze 29, 11
        add 11, 11, 6
        mullw 12, 3, 5
        addc 8, 12, 8
        mulhwu 9, 3, 5
        addze 9, 9
        addc 9, 29, 9
        addze 4, 4
        lwz 29, 20(1)                           # 4-byte Folded Reload
        mulhwu 30, 10, 5
        mullw 10, 10, 5
        addc 9, 10, 9
        adde 4, 30, 4
        srawi 5, 5, 31
        add 10, 11, 10
        mullw 12, 5, 3
        lwz 30, 24(1)                           # 4-byte Folded Reload
        mulhw 3, 5, 3
        addc 5, 12, 6
        adde 3, 3, 10
        addc 5, 9, 5
        adde 3, 4, 3
        srawi 4, 8, 31
        srawi 6, 0, 31
        xor 3, 3, 4
        xor 4, 5, 4
        or 3, 4, 3
        xor 4, 8, 6
        or 3, 3, 4
        cntlzw  3, 3
        not     3, 3
        rlwinm 3, 3, 27, 31, 31
        addi 1, 1, 32
        blr
.Lfunc_end0:
        .size   no__mulodi4, .Lfunc_end0-.Lfunc_begin0
        .cfi_endproc
```

**Comment by tstellar** [comment]
@topperc What do you think about backporting this? https://github.com/llvm/llvm-project/commit/c1a31ee65b3a2bf2b452febb99703b3fdfa43328

**Comment by topperc** [comment]
> @topperc What do you think about backporting this? [c1a31ee](https://github.com/llvm/llvm-project/commit/c1a31ee65b3a2bf2b452febb99703b3fdfa43328)

Looks ok for backport.

**Comment by tstellar** [comment]
Merged: add3ab7f4c8a
